opm/simulators/linalg/bda/fgpilu.cpp

```diff
@@ -0,0 +1,570 @@
+            d_L_cols = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * L_nnzbs);
+            d_LU_cols = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(int) * LU_nnzbs);
+            d_Ltmp = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(double) * L_nnzbs * block_size * block_size);
+            d_Utmp = cl::Buffer(*context, CL_MEM_READ_WRITE, sizeof(double) * Ut_nnzbs * block_size * block_size);
+
+            Dune::Timer t_copy_pattern;
+            err |= queue->enqueueWriteBuffer(d_Ut_ptrs, CL_TRUE, 0, sizeof(int) * (Nb+1), Ut_ptrs);
+            err |= queue->enqueueWriteBuffer(d_L_rows, CL_TRUE, 0, sizeof(int) * (Nb+1), L_rows);
+            err |= queue->enqueueWriteBuffer(d_LU_rows, CL_TRUE, 0, sizeof(int) * (Nb+1), LU_rows);
+            err |= queue->enqueueWriteBuffer(d_Ut_idxs, CL_TRUE, 0, sizeof(int) * Ut_nnzbs, Ut_idxs);
+            err |= queue->enqueueWriteBuffer(d_L_cols, CL_TRUE, 0, sizeof(int) * L_nnzbs, L_cols);
+            err |= queue->enqueueWriteBuffer(d_LU_cols, CL_TRUE, 0, sizeof(int) * LU_nnzbs, LU_cols);
+            if (verbosity >= 3){
+                std::ostringstream out;
+                out << "FGPILU copy sparsity pattern time: " << t_copy_pattern.stop() << " s";
+                OpmLog::info(out.str());
+            }
+            if (verbosity >= 2){
+                std::ostringstream out;
+                out << "FGPILU PARALLEL: " << PARALLEL;
+                OpmLog::info(out.str());
+            }
+
+            initialized = true;
+        }
+
+
+        // copy to GPU
+        Dune::Timer t_copy1;
+        err = queue->enqueueWriteBuffer(d_Ut_vals, CL_TRUE, 0, sizeof(double) * Ut_nnzbs * block_size * block_size, Ut_vals);
+        err |= queue->enqueueWriteBuffer(d_L_vals, CL_TRUE, 0, sizeof(double) * L_nnzbs * block_size * block_size, L_vals);
+        err |= queue->enqueueWriteBuffer(d_LU_vals, CL_TRUE, 0, sizeof(double) * LU_nnzbs * block_size * block_size, LU_vals);
+        if (verbosity >= 3){
+            std::ostringstream out;
+            out << "FGPILU copy1 time: " << t_copy1.stop() << " s";
+            OpmLog::info(out.str());
+        }
+        if (err != CL_SUCCESS) {
+            // enqueueWriteBuffer is C and does not throw exceptions like C++ OpenCL
+            OPM_THROW(std::logic_error, "FGPILU OpenCL enqueueWriteBuffer error");
+        }
+
+        // call kernel
+        for (int sweep = 0; sweep < num_sweeps; ++sweep) {
+            // normally, L_vals and Ltmp are swapped after the sweep is done
+            // these conditionals implement that without actually swapping pointers
+            // 1st sweep reads X_vals, writes to Xtmp
+            // 2nd sweep reads Xtmp, writes to X_vals
+            auto *Larg1 = (sweep % 2 == 0) ? &d_L_vals : &d_Ltmp;
+            auto *Larg2 = (sweep % 2 == 0) ? &d_Ltmp : &d_L_vals;
+            auto *Uarg1 = (sweep % 2 == 0) ? &d_Ut_vals : &d_Utmp;
+            auto *Uarg2 = (sweep % 2 == 0) ? &d_Utmp : &d_Ut_vals;
```

Comment:
Hence we are basically doing what I would call a Jacobi sweep (neglecting all new information in other rows). 

In the future we might want to test what happens for our problems if we allow to use the new information, similar to [this paper](https://link.springer.com/chapter/10.1007/978-3-662-48096-0_50)

Reply:
It is indeed a Jacobi sweep, but using new information in other rows introduces datadependencies which the exact ILU decomposition also has. It is probably better to use exact decomposition in that case.
That paper talks about iterative triangular solves, which could replace the ILU_apply part, but I think it is less relevant for this PR (although we could skip reversing the rows of U in that case).